src/entities/pools/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,226 @@
+import { createPublicClient, encodeFunctionData, http } from 'viem';
+import {
+    BaseJoin,
+    JoinCallInput,
+    JoinInput,
+    JoinQueryResult,
+    PoolState,
+} from '..';
+import { Address } from '../../../types';
+import {
+    BALANCER_HELPERS,
+    BALANCER_VAULT,
+    CHAINS,
+    ZERO_ADDRESS,
+} from '../../../utils';
+import { WeightedEncoder } from '../../encoders/weighted';
+import { TokenAmount } from '../../tokenAmount';
+import { balancerHelpersAbi } from '../../../abi/balancerHelpers';
+import { Token } from '../../token';
+import { vaultAbi } from '../../../abi';
+
+export class WeightedJoin implements BaseJoin {
+    // TODO - Probably not needed
+    getInstance(): WeightedJoin {
+        return new WeightedJoin();
+    }
+
+    public async query(
+        input: JoinInput,
+        poolState: PoolState,
+    ): Promise<JoinQueryResult> {
+        // TODO - This would need extended to work with relayer
+
+        this.checkInputs(input, poolState);
+
+        // infer join kind by input tokens
+        const joinKind = this.getJoinKind(input, poolState.address);
+
+        // Initialize join parameters
+        let maxAmountsIn = Array(poolState.tokens.length).fill(0n);
+        let userData: Address;
+        const poolAssets = poolState.tokens.map((t) => t.address);
+
+        switch (joinKind) {
+            case 'Init': {
+                maxAmountsIn = this.getAmountsIn(input, poolAssets);
+                userData = WeightedEncoder.joinInit(maxAmountsIn);
+                break;
+            }
+            case 'GivenIn': {
```

Comment:
imo we should take this opportunity as a chance to get away from these hard to reason about name. I would lean more towards mapping to things like `proportional`, `unbalanced`, `singleAsset`, etc

Reply:
I get your point of renaming to make the interface more intuitive. That was the goal of replacing `ExactTokensInForBPTOut` by `GivenIn` 😅 but I agree that this can be improved.

Just trying to understand what you mean by differentiating `proportional` and `unbalanced` though. Is that in the sense of having some sort of helper that would automatically calculate proportional amounts in based on a single asset amount in as reference? I can see that being helpful, but just wanted to make sure this was the suggestion.

And I'd love some help on naming the `givenOut` alternative - maybe `exactOut`?

Reply:
I'd probably do something like `SingleAssetExactAmountIn` `SingleAssetExactBptOut`

Reply:
Regarding naming, I was trying to move away from being that verbose, but I’m having a hard time thinking about alternatives.
I did find those old `ExactTokensInForTokenOut` pretty confusing, but thinking a bit more about it I’d say the confusion came from having `TokenIn` and `TokenOut` in the same name. I’d always have to read a few times to make sure I was using the correct function.
Having said that, when I read `SingleAssetJoin` and `ExactBptOut` in the same name, I kinda expect it to perform a single asset join, but with an exact amount out. Which is not the case, right? I's actually performing a `proportional` join with an exact amount out.
In any case, I'd still consider a `ProportionalExactAmountsInJoinInput` + `ProportionalExactBptOutJoinInput` not as good as simply `ProportionalJoinInput` + `ExactBptOutJoinInput`.
In fact, I'd actually go for `ExactAmounOutJoinInput` - which seems easier to understand than `Bpt` for devs that are not super into Balancer stuff.
Let me know what you think

Reply:
After talking to Daniel, I realized that my understanding of the request of moving away from current names was different from what he was expecting.
I thought the idea was to rethink the whole interface towards the user interaction and then map those options down to the contract interface.
For example, we'd have `proportional` + `unbalanced` + `singleAsset` on the SDK interface, with different inputs and logic, but they would all map to down to `EXACT_TOKENS_IN_FOR_BPT_OUT` in the contract interface.
That was not his intention. He wanted to simply rename the 4 options we currently have on the contract interface to something easier to read/remember and also more descriptive. For example:
- `INIT` -> `Init`
- `EXACT_TOKENS_IN_FOR_BPT_OUT` -> `Unbalanced`
- `TOKEN_IN_FOR_EXACT_BPT_OUT` -> `SingleAsset`
- `ALL_TOKENS_IN_FOR_EXACT_BPT_OUT` -> `Proportional`